src/functional/ffront/func_to_foast.py

```diff
@@ -133,35 +261,79 @@ def apply(
-                raise ValueError(
-                    "Can not create field operator from a function that is not in a source file!"
-                )
-            return textwrap.dedent(inspect.getsource(func))
+    def apply_to_function(
+        cls,
+        func: types.FunctionType,
+        externals: Optional[dict[str, Any]] = None,
+    ) -> foast.FieldOperator:
+        source_definition = SourceDefinition.from_function(func)
+        closure_refs = ClosureRefs.from_function(func)
+        return cls.apply(source_definition, closure_refs, externals)
 
-        source = get_src(func)
-        return cls.apply(
-            source, filename=inspect.getabsfile(func), starting_line=inspect.getsourcelines(func)[1]
+    def visit_FunctionDef(self, node: ast.FunctionDef, **kwargs) -> foast.FieldOperator:
+        _, _, imported_names, nonlocal_names, global_names = SymbolNames.from_source(
+            self.source, self.filename
         )
+        if missing_defs := (self.closure_refs.unbound - imported_names):
+            raise common.GTValueError(f"Missing reference definitions: {missing_defs}")
+
+        # 'SymbolNames.from_source()' uses the symtable module to analyze the isolated source
+        # code of the function, and thus all non-local symbols are classified as 'global'.
+        # However, 'closure_refs' comes from inspecting the live function object, which might
+        # have not been defined at a global scope, and therefore actual symbol values could appear
+        # in both 'closure_refs.globals' and 'self.closure_refs.nonlocals'.
+        defs = self.closure_refs.globals | self.closure_refs.nonlocals
+        closure = [
+            symbol_makers.make_symbol_from_value(
+                name, defs[name], foast.Namespace.CLOSURE, self._make_loc(node)
+            )
+            for name in global_names | nonlocal_names
+        ]
 
-    def visit_FunctionDef(self, node: ast.FunctionDef, **kwargs) -> foast.FieldOperator:
         return foast.FieldOperator(
             id=node.name,
             params=self.visit(node.args),
             body=self.visit_stmt_list(node.body),
-            location=self._getloc(node),
+            closure=closure,
+            location=self._make_loc(node),
+        )
+
+    def visit_Import(self, node: ast.Import, **kwargs) -> None:
+        self._make_syntax_error(
+            node, f"Only 'from' imports from {fbuiltins.MODULE_BUILTIN_NAMES} are supported"
         )
 
+    def visit_ImportFrom(self, node: ast.ImportFrom, **kwargs) -> None:
+        if node.module not in fbuiltins.MODULE_BUILTIN_NAMES:
+            self._make_syntax_error(
+                node, f"Only 'from' imports from {fbuiltins.MODULE_BUILTIN_NAMES} are supported"
+            )
+
+        symbols = []
+        for alias in node.names:
+            if alias.name not in self.externals_defs:
+                raise self._make_syntax_error(
+                    node, message="Missing imported symbol '{alias.name}'"
```

Comment:
Do I understand this correctly: this error message is triggered when the user forgets to supply `X` in the externals and then tries to `from externals import X`?

Reply:
Yes, that's correct.